CenterForOpenScience/osf.io
# Feature/Optimize Meetings Page Serialization [PLAT-657]

# Purpose

The serialization of meetings pages (`/view/<meeting>/` route) is incredibly slow and expensive. For example, SPSP 2016 has somewhere around 1,000 submissions that are all serialized in this response.

# Changes

Almost all speed gains are attained from adding pagination.  Instead of potentially 1000+ records being serialized before the page loads, 50 at most are returned.  Because treebeard assumes that all records are loaded for sorting and filtering, (which is no longer true with pagination), custom sorting and filtering were added. UI should look and behave very similarly, details below on changes.  Page loads much more quickly at the expense of losing instantaneous filtering/sorting.

### Pagination
- Max fifty records loaded at a time instead of everything.
- Pagination added at bottom of the page - UI should look and behave identically to pagination in our ember apps (This is new!)
- Every time you paginate, sort, or filter, a new request is sent to the server, and that query is handled server-side (new, previously was handled client-side).

### Sorting/Filtering
- Sort headings/filter bar pulled from treebeard and built myself.  Some column widths have shifted *slightly*.
- Loads with a default sorting on download count, highest to lowest.  Need default sorting for reliable pagination (pretty sure previous setup had no default sort).
- Can still sort by title, author, downloads, created, and category (same as before).
- Filtering is no longer onkeyup, (since not doing this client-side). Type in your filter, and press enter and a request will be sent.
- Some raw sql and extra() had to be used for the time-being for filtering/sorting on first visible contributor.  (Avoided this ProgrammingError: subquery must return only one column), which according to @caseyrollins  "right now the auto including breaks subqueries where both halves of the query are Guid objects" (QA can ignore this)
- Filtering looks through project title and first visible contributor fullname (previously looking through family name if exists, otherwise fullname)
- Sorting authors sorts on first visible contributor family name (previously sorted on family name unless none given, then sorted on fullname)
- Sort/Filter query parameters retained on pagination
- On filtering, sort parameter retained, but page is set to 1
- On sorting, filtering parameter retained, but page is set to 1

![screen shot 2018-02-28 at 2 34 17 pm](https://user-images.githubusercontent.com/9755598/36811846-d3bb0cb0-1c94-11e8-874b-5deb46199027.png)
![giphy 3](https://user-images.githubusercontent.com/9755598/37062869-4f8d8706-215d-11e8-8e91-8582967944f4.gif)



# QA
- **NO PREP NEEDED BEFOREHAND**
- Meetings page should behave very similarly, see "Changes" for details.  Some things have had to change due to pagination being added.
- Make sure pagination, sorting, filtering works, and query parameters are retained, or switched to 1, when necessary
- Test some meetings that have different ranges of numbers of submissions, to test pagination logic at bottom doesn't behave weirdly (it shouldn't!)

# Ticket
https://openscience.atlassian.net/browse/PLAT-657
